File: src/main/java/com/akagiyui/drive/model/response/RoleInfoResponse.java
package com.akagiyui.drive.model.response;

import com.akagiyui.drive.entity.Role;
import lombok.Data;
import lombok.experimental.Accessors;

import java.util.List;

/**
 * 角色信息响应
 *
 * @author AkagiYui
 */
@Data
@Accessors(chain = true)
public class RoleInfoResponse {

    /**
     * 角色id
     */
    private String id;

    /**
     * 角色名
     */
    private String name;

    /**
     * 角色描述
     */
    private String description;

    /**
     * 角色是否被禁用
     */
    private boolean disabled;

    /**
     * 是否是默认角色
     */
    private boolean isDefault;

    /**
     * 角色用户数
     */
    private long userCount;

    /**
     * 角色权限
     */
    private List<String> permissions;

    public RoleInfoResponse(Role role) {
        this.id = role.getId();
        this.name = role.getName();
        this.description = role.getDescription();
        this.disabled = role.getDisabled();
        this.isDefault = role.getIsDefault();
        this.userCount = role.getUsers().size();
        this.permissions = role.getPermissions().stream().map(Enum::name).toList();
    }

    /**
     * 从角色列表转换
     *
     * @param roles 角色列表
     * @return 角色信息响应列表
     */
    public static List<RoleInfoResponse> fromRoleList(List<Role> roles) {
        return roles.stream().map(RoleInfoResponse::new).toList();
    }

}

File: app/src/main/kotlin/com/akagiyui/drive/model/response/RoleInfoResponse.kt
package com.akagiyui.drive.model.response

import com.akagiyui.drive.entity.Role

/**
 * 角色信息响应
 *
 * @author AkagiYui
 */
data class RoleInfoResponse(
    /**
     * 角色ID
     */
    val id: String,

    /**
     * 角色名
     */
    val name: String,

    /**
     * 角色描述
     */
    val description: String,

    /**
     * 角色是否被禁用
     */
    val disabled: Boolean,

    /**
     * 是否是默认角色
     */
    val isDefault: Boolean,

    /**
     * 角色用户数
     */
    val userCount: Long,

    /**
     * 角色权限
     */
    val permissions: List<String>,
) {
    constructor(role: Role) : this(
        id = role.id,
        name = role.name,
        description = role.description ?: "",
        disabled = role.disabled,
        isDefault = role.isDefault,
        userCount = role.users.size.toLong(),
        permissions = role.permissions.map { it.name }
    )
}

/**
 * 从角色列表转换
 *
 * @return 角色信息响应列表
 */
fun List<Role>.toResponse(): List<RoleInfoResponse> {
    return this.map { RoleInfoResponse(it) }.toList()
}
